libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/RouteProcessorRunnable.kt

```diff
@@ -0,0 +1,180 @@
+package com.mapbox.services.android.navigation.v5.internal.navigation
+
+import android.location.Location
+import android.os.Handler
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.navigator.NavigationStatus
+import com.mapbox.navigator.RouteState
+import com.mapbox.services.android.navigation.v5.milestone.Milestone
+import com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation
+import com.mapbox.services.android.navigation.v5.navigation.MapboxNavigationOptions
+import com.mapbox.services.android.navigation.v5.offroute.OffRouteDetector
+import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress
+import com.mapbox.services.android.navigation.v5.snap.SnapToRoute
+import java.util.ArrayList
+import java.util.Date
+
+internal class RouteProcessorRunnable(
+    private val routeProcessor: NavigationRouteProcessor,
+    private val navigation: MapboxNavigation,
+    private val workerHandler: Handler,
+    private val responseHandler: Handler,
+    private val listener: RouteProcessorBackgroundThread.Listener
+) : Runnable {
+    private lateinit var rawLocation: Location
+
+    companion object {
+        private const val ONE_SECOND_IN_MILLISECONDS = 1000
+        private const val ARRIVAL_ZONE_RADIUS = 40
+    }
+
+    override fun run() {
+        if (::rawLocation.isInitialized) {
+            process()
+        }
+    }
+
+    fun updateRawLocation(rawLocation: Location) {
+        this.rawLocation = rawLocation
+    }
+
+    private fun process() {
+        val mapboxNavigator = navigation.retrieveMapboxNavigator()
+        val options = navigation.options()
+        val route = navigation.route
+
+        val date = Date()
+        var status = mapboxNavigator.retrieveStatus(
+            date,
+            options.navigationLocationEngineIntervalLagInMilliseconds().toLong()
+        )
+        val previousStatus = routeProcessor.retrievePreviousStatus()
```

Comment:
since this method returns nullable does it make sense to declare ref as `val previousStatus: NavigationStatus? = ....`?